enhancements/cloud-integration/azure/azure-workload-identity.md

```diff
@@ -0,0 +1,401 @@
+  azure_tenant_id: <tenant id>
+kind: Secret
+type: Opaque
+```
+
+We propose that when utilizing Azure Workload Identity, the credentials secret will contain an `azure_client_id` that is the `clientID` of the Managed Identity provisioned by `ccotcl` for the operator. The `azure_client_secret` key will be absent and instead we can provide the path to the mounted `ServiceAccount` token as an `azure_federated_token_file` key; the path to the mounted token is well
+known and is specified in the operator deployment.
+
+The resource group in which the installer will create infrastructure will not be known when these secrets are generated by `ccoctl` ahead of installation and operators which rely on `azure_resourcegroup` and `azure_resource_prefix` such as the
+[image-registry](https://github.com/openshift/cluster-image-registry-operator/blob/8556fd48027f89e19daad36e280b60eb93d012d4/pkg/storage/azure/azure.go#L95-L100) should obtain the resource group details from the cluster `Infrastructure` object instead.
+
+```yaml
+apiVersion: v1
+data:
+  azure_client_id: <client id>
+  azure_federated_token_file: <path to mounted service account token, eg. "/var/run/secrets/openshift/serviceaccount/token">
+  azure_region: <region>
+  azure_subscription_id: <subscription id>
+  azure_tenant_id: <tenant id>
+kind: Secret
+type: Opaque
+```
+
+#### Creating workload identity clients in operators
+
+In order to create Azure clients which utilize a `ClientAssertionCredential`, operators must update to version `>= v1.2.0` of the azidentity package within [azure-sdk-for-go](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity@v1.2.0). Ahead of this work, due to the [end of life
+announcement](https://techcommunity.microsoft.com/t5/microsoft-entra-azure-ad-blog/microsoft-entra-change-announcements-september-2022-train/ba-p/2967454) of the Azure Active Directory Authentication Library (ADAL), PRs (ex. [openshift/cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/pull/846)) have been opened for operators to migrate to creating clients via
+azidentity which are converted into an authorizer for use with v1 clients. Once these changes have been made, we propose that OpenShift operators continue to utilize a config secret to obtain authentication details as described in the previous section but create workload identity clients when the `azure_client_secret` is absent AND/OR when  `azure_federated_token_file` fields are found in the
+config. Config secrets will be generated by cluster creators prior to installation by using `ccoctl` and will be provided as manifests for install.
+
+Due to the deployment of the Azure Workload Identity mutating admission webhook, environment variables should also be respected by client instatiation as an alternative way of supplying the `clientID` eg. `AZURE_CLIENT_ID`, `tenantID` eg. `AZURE_TENANT_ID` and `federatedTokenFile` eg. `AZURE_FEDERATED_TOKEN_FILE`.
```

Comment:
```suggestion
Due to the deployment of the Azure Workload Identity mutating admission webhook, environment variables should also be respected by client instantiation as an alternative way of supplying the `clientID` eg. `AZURE_CLIENT_ID`, `tenantID` eg. `AZURE_TENANT_ID` and `federatedTokenFile` eg. `AZURE_FEDERATED_TOKEN_FILE`.
```

Sorry, does this mean that the functions from the azidentity library that the OpenShift operator uses to instantiate the client will check for those env vars internally without the operator code needing to do anything; or does this mean that the OpenShift code instantiating the client needs to be aware of the env vars so they can be (massaged and?) passed into the client instantiation functions?

[Later] Thought the code sample below might shed some light, but it doesn't reference env vars at all. Which could mean that the `azidentity.NewClientAssertionCredential()` function internally uses them (in which case, what are the precedence rules?) or that the code isn't taking the env vars into account (yet?).

Another question: If the mutating webhook is going to be a thing, why do we still need the config object?
OR
If we can provide the information we need via the config object, why do we need the webhook? (Um, I guess the answer to this one is that the webhook is the only way we get the volume with the token file?)

Reply:
The azidentity library will not check for the environment variables currently but [may check for them in the future](https://github.com/Azure/azure-sdk-for-go/issues/15615#issuecomment-1211012677). Using the environment variables will come from our implementation. I'm calling it out here due to the mutating webhook setting these.

> Another question: If the mutating webhook is going to be a thing, why do we still need the config object?
OR
If we can provide the information we need via the config object, why do we need the webhook? (Um, I guess the answer to this one is that the webhook is the only way we get the volume with the token file?)

The webhook will not be needed for our operators which have config secrets to get the necessary information from. If we wanted to use the webhook to set the environment variables for our operator pods then we would need operators (or something) to annotate their `ServiceAccounts` with the `clientID` & `tenantID`. I don't know how feasible that would be and whatever is annotating the `ServiceAccount` would need to get the information from somewhere.